src/btree/row_modify.c

```diff
@@ -403,22 +402,14 @@ __wt_update_obsolete_check(
             continue;
 
         /*
-         * WiredTiger internal operations such as rollback to stable and prepare transaction
-         * rollback adds a globally visible tombstone to the update chain to remove the entire key.
-         * Treating these globally visible tombstones as obsolete and trimming update list can cause
-         * problems if the update chain is getting accessed somewhere else. To avoid this problem,
-         * skip these globally visible tombstones from the update obsolete check that were generated
-         * from prepare transaction rollback but not from RTS, because there are no concurrent
-         * operations run in parallel to the RTS to be affected.
+         * Prepare transaction rollback adds a globally visible tombstone to the update chain to
+         * remove the entire key. Treating these globally visible tombstones as obsolete and
+         * trimming update list can cause problems if the update chain is getting accessed somewhere
+         * else. To avoid this problem, skip these globally visible tombstones from the update
+         * obsolete check.
          */
-        if (upd->txnid == WT_TXN_NONE && upd->upd_start_ts == WT_TS_NONE &&
-          upd->type == WT_UPDATE_TOMBSTONE && upd->next != NULL &&
-          upd->next->txnid == WT_TXN_ABORTED) {
-            WT_ACQUIRE_READ(prepare_state, upd->next->prepare_state);
-            /* We may see a locked prepare state if we race with prepare rollback. */
-            if (prepare_state == WT_PREPARE_INPROGRESS || prepare_state == WT_PREPARE_LOCKED)
-                continue;
-        }
+        if (F_ISSET(upd, WT_UPDATE_PREPARE_ROLLBACK))
```

Comment:
Do we need to check for the prepared updates specifically? The main reason is that in the below code, if the tree is marked for garbage collect, we didn't check whether it is prepared or not?

Reply:
no, it must be prepared in this case if we add this tombstone. I can add an assert for that if you want.

Reply:
We are passing only the next update of the currently inserted update, so there is no way a prepare can be there.